source/common/formatter/substitution_formatter.cc

```diff
@@ -205,8 +212,12 @@ StructFormatter::toFormatListValue(const ProtobufWkt::ListValue& list_value_form
 
 std::vector<FormatterProviderPtr>
 StructFormatter::toFormatStringValue(const std::string& string_format) const {
+  if (commands_.has_value()) {
+    return SubstitutionFormatParser::parse(string_format, commands_.value());
+  }
+
   return SubstitutionFormatParser::parse(string_format);
-};
+}
```

Comment:
How about we just store the commands as a regular vector? `std::vector<FormatterProviderPtr> SubstitutionFormatParser::parse(const std::string& format)` just passes an empty vec to the other overload, so seems like this would simplify things a bit?

Reply:
See my comment from above, I went out of my way to avoid copying that vector because really you only needed at construction time... what's special about `StructFormatter`'s constructor is that it has recursive calls to handle the JSON (nested) doc.... Thus we keep around the reference to avoid passing it around the call chain.

Reply:
I don't think this affects copying the vector, its just a question about whether you set the vector to empty vec (no copying) or absl::empty (no copying).

Alternatively you could do `SubstitutionFormatParser::parse(string_format, commands_.value_or(<empty vector>);`